[[ Bug 20027 ]] Ensure UQL is cleard when setting container value

montegoulding opened the pull request:
This patch ensures that undeclared variables are cleared of their UQL
status when their value is set via `MCContainer::set_valueref`.

montegoulding commented on tests/lcs/core/execution/uql.livecodescript:
```diff
@@ -0,0 +1,36 @@
+	put "foo" after tTest2
+	TestAssert "unquoted literal put after", tTest2 is "foo"
+	put "foo" before tTest3
+	TestAssert "unquoted literal put before", tTest3 is "foo"
```
@runrevmark when you get a chance could you review these tests as I don't recall when the change was made (if it was deliberate) for UQLs to be cleared of their value in `put before|after` situations. Previously if doing something like appending to a variable you were required to `put empty into var` before you started appending to it otherwise you got the var name included in the value.

As in:
```
put "foo" after bar
-- bar is now "barfoo"
```
But now:
```
put "foo" after bar
-- bar is now "foo"
```

runrevmark replied:
It would have been a long time ago - certainly 7.0 at the earliest...

Indeed https://quality.livecode.com/show_bug.cgi?id=14536 was fixed in a 7 release and it basically put the clear_uql() call into the matchtext/binaryDecode which seems to have disappeared since (where you have put it is a more sensible place).

My guess is that when we containerized variables the refactoring missed the need for clearuql in those two places.

runrevmark replied:
This goes back to before 6.7.11 as well - MCVarref's used to do a clearuql before they were accessed - so *not* clearing them was probably fixed a long time ago.

runrevmark commented:
There are three other places which need to be tested here:
  - binaryDecode
  - queryRegistry
  - setting externalV1 variable refs

It also needs to be fixed for the SetVariable (and probably friend) calls in externalV0 interface - at the moment the latter is using MCContainer::set - but it could just use set_valueref instead (thus meaning the same code as all the other places is used).

runrevmark commented:
Note: I've set the milestone to 9.0.5 for now - but there might be a bit too much in the above for it to be done in time; so it might slip to 9.0.6.

montegoulding committed:
[[ Bug 20027 ]] Ensure UQL is cleared when setting container value

This patch ensures that undeclared variables are cleared of their UQL
status when their value is set via `MCContainer::set_valueref`.

montegoulding commented:
@runrevmark I have added tests for binaryDecode and queryRegistry but the external tests will need a bit of work to put together an external test harness I think.

runrevmark commented:
@montegoulding Okay - could you file a bug about needing a test harness for both the external interfaces and make sure it links to this bug as one case which needs an explicit case (so we don't forget).

runrevmark commented:
@livecode-vulcan review ok a2ae215

livecode-vulcan commented:
💙 review by @runrevmark ok a2ae215d5977230481df3562cb21215db18a2424

livecode-vulcan commented:
😎 test success a2ae215d5977230481df3562cb21215db18a2424

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/517)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/509)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/509)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/509)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/509)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2031)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/259)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/258)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2063)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2053)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2051)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2086)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2023)

montegoulding commented:
https://quality.livecode.com/show_bug.cgi?id=22059
